src/ws.nim

```diff
@@ -0,0 +1,415 @@
+
+  # Payload length can be 7 bits, 7+16 bits, or 7+64 bits.
+  var finalLen: uint64 = 0
+
+  let headerLen = uint(b1 and 0x7f)
+  if headerLen == 0x7e:
+    # Length must be 7+16 bits.
+    var length = newSeq[byte](2)
+    await ws.tcpSocket.readExactly(addr length[0], 2)
+    finalLen = cast[ptr uint16](length[0].addr)[].toBE
+  elif headerLen == 0x7f:
+    # Length must be 7+64 bits.
+    var length = newSeq[byte](8)
+    await ws.tcpSocket.readExactly(addr length[0], 8)
+    finalLen = cast[ptr uint64](length[0].addr)[].toBE
+  else:
+    # Length must be 7 bits.
+    finalLen = headerLen
+
+  # Do we need to apply mask?
+  frame.mask = (b1 and 0x80) == 0x80
+
+  if ws.masked == frame.mask:
+    # Server sends unmasked but accepts only masked.
+    # Client sends masked but accepts only unmasked.
+    raise newException(WebSocketError, "Socket mask mismatch")
+
+  var maskKey = newSeq[byte](4)
+  if frame.mask:
+    # Read the mask.
+    await ws.tcpSocket.readExactly(addr maskKey[0], 4)
+
+  # Read the data.
+  var data = newSeq[byte](finalLen)
+  await ws.tcpSocket.readExactly(addr data[0], int finalLen)
+  frame.data = data
+  if frame.data.len != int finalLen:
+    raise newException(WebSocketError, "Failed to read websocket frame data")
+
+  if frame.mask:
+    # Apply mask, if we need too.
+    for i in 0 ..< frame.data.len:
+      frame.data[i] = (frame.data[i].uint8 xor maskKey[i mod 4].uint8)
+  return frame
+
+proc receivePacket*(ws: WebSocket): Future[(Opcode, seq[byte])] {.async.} =
+  ## Wait for a string or binary packet to come in.
+  var frame = await ws.receiveFrame()
+
+  var packet = frame.data
+  # If there are more parts read and wait for them
+  while frame.fin != true:
+    frame = await ws.receiveFrame()
+    if frame.opcode != Cont:
+      raise newException(WebSocketError, "Expected continuation frame")
+    packet.add frame.data
+  return (frame.opcode, packet)
+
+proc receiveStrPacket*(ws: WebSocket): Future[seq[byte]] {.async.} =
+  ## Wait only for only string and control packet to come. Errors out on Binary packets.
+  let (opcode, data) = await ws.receivePacket()
+  case opcode:
+    of Text:
+      return data
+    of Binary:
+      raise newException(WebSocketError, "Expected string packet, received binary packet")
+    of Ping:
+      await ws.send(data, Pong)
+    of Pong:
+      discard
+    of Cont:
+      discard
+    of Close:
+      ws.close()
+
+proc validateWSClientHandshake*(transp: StreamTransport,
+    header: HttpResponseHeader): void =
+  if header.code != ord(Http101):
+    raise newException(WebSocketError, "Server did not reply with a websocket upgrade: " &
+          "Header code: " & $header.code &
+          "Header reason: " & header.reason() &
+          "Address: " & $transp.remoteAddress())
+
+proc newWebsocketClient*(uri: Uri, protocols: seq[string] = @[]): Future[
+    WebSocket] {.async.} =
+  let
+    keyDec = align(
```

Comment:
**SECURITY WARNING:** We are using epoch seconds as security key? Is this from WS specification, or just proper random number could be used?

Reply:
Done